plugin/pulp/plugin/download.py

```diff
@@ -0,0 +1,122 @@
+"""
+This module contains set of download APIs.
+"""
+import os
+import hashlib
+
+import requests
+from requests_futures.sessions import FuturesSession
+from requests.packages.urllib3.util import retry
+
+# TODO make config settings
+DEFAULT_CONNECT_TIMEOUT = 6.1
+DEFAULT_READ_TIMEOUT = 30
+DEFAULT_TRIES = 5
+DEFAULT_CONCURRENCY = 5
+# TODO Picked a bit out of a hat - I expect it should try to line up with the FS.
+DEFAULT_CHUNKSIZE = 1024 * 1024
+
+
+def default_callback_generator(filename=None, size=None, checksums=None):
+    """
+    Creates a closure to write a response to disk, checksum it, and verify
+    its size.
+
+    :param filename: The absolute path the response should be saved to.
+    :param size: The size, in bytes, that the file is expected to be; check is
+                 skipped if this value is not provided.
+    :param checksums: A dictionary of checksums to validate the file with; the
+                      key should be the hash algorithm and the value the
+                      expected hex digest.
+    """
+    # TODO potentially add a way to have this generate the entire checksum set.
+    # Maybe if there's {'algorithm': None}?
+    checksums = checksums or {}
+
+    def callback(self, response):
+        """
+        A callback for the requests_futures library.
+
+        :param resp: The requests response object
+        """
+        hashes = {algorithm: hashlib.new(algorithm) for algorithm in checksums}
+        with open(filename, mode='wb') as fp:
+            chunks = response.iter_content(chunk_size=DEFAULT_CHUNKSIZE)
+            for chunk in chunks:
+                for hash in hashes.values():
```

Comment:
If we're worried about stomping builtins, `import builtins; builtins.hash` so it's explicit where they're coming from. The python `hash` builtin has no use in this context, and the `hash` identifier here is explicit and clear.